# Imports

In [1]:
import pandas as pd
import numpy as np

***

# Carga de datos

In [3]:
auctions = pd.read_pickle("../../data/tp2/auctions_tp2_formateado.pkl")
clicks = pd.read_pickle("../../data/tp2/clicks_tp2_formateado.pkl")
events = pd.read_pickle("../../data/tp2/events_tp2_formateado.pkl")
installs = pd.read_pickle("../../data/tp2/installs_tp2_formateado.pkl")
targets = pd.read_pickle("../../data/tp2/ref_hashes_target.pkl")

***

# Inspección de columnas y valores

In [51]:
auctions.head(5).T

10129864                    25285906  \
date         2019-04-18 00:00:00.015050  2019-04-18 00:00:00.029014   
device_id           1826643666390887030         7037174172278258682   
ref_type_id                           7                           1   
source_id                             0                           0   

                               3946062                     3946063   \
date         2019-04-18 00:00:00.057540  2019-04-18 00:00:00.126828   
device_id           3392065368947589877         1228982273563226229   
ref_type_id                           1                           1   
source_id                             1                           1   

                               24607726  
date         2019-04-18 00:00:00.132510  
device_id           4123059034628125459  
ref_type_id                           1  
source_id                             8

In [52]:
clicks.head(5).T

0  \
advertiser_id                                   1   
action_id                                     NaN   
source_id                                       2   
created          2019-04-18 05:27:42.197000+00:00   
latitude                                  1.71455   
longitude                                0.871535   
wifi_connection                             False   
carrier_id                                      3   
trans_id          9JMAfrb-b9cSEVCJb0P9JfihGthaS7E   
os_minor                              1.51764e+18   
agent_device                                  NaN   
os_major                              5.13162e+18   
specs_brand                     71913840936116953   
brand                                           0   
timeToClick                                 2.317   
touchX                                      0.968   
touchY                                      0.503   
ref_hash                      1293710398598742392   
os_es_android                                   1   

                                                1  \
advertiser_id                                   1   
action_id                                     NaN   
source_id                                       1   
created          2019-04-18 05:27:03.164000+00:00   
latitude                                  1.71451   
longitude                                0.871062   
wifi_connection                              True   
carrier_id                                      2   
trans_id          r3xtTRv2lInfiXG8JI3NQsNcBo8GyFQ   
os_minor                              1.28858e+18   
agent_device                                  NaN   
os_major                              3.90839e+18   
specs_brand                   3576558787748411622   
brand                                           1   
timeToClick                                 7.653   
touchX                                      0.712   
touchY                                      1.689   
ref_hash                      1663930990551616564   
os_es_android                                   1   

                                                2  \
advertiser_id                                   1   
action_id                                     NaN   
source_id                                       1   
created          2019-04-18 05:42:07.926000+00:00   
latitude                                  1.71455   
longitude                                0.871535   
wifi_connection                              True   
carrier_id                                      4   
trans_id          WOnHFqQtY48z_ygKZ-030U_g0TMGVMw   
os_minor                              2.23874e+18   
agent_device                                  NaN   
os_major                              3.58123e+18   
specs_brand                   3576558787748411622   
brand                                         NaN   
timeToClick                               464.796   
touchX                                      0.227   
touchY                                      0.251   
ref_hash                      8488038938665586188   
os_es_android                                   1   

                                                3  \
advertiser_id                                   1   
action_id                                     NaN   
source_id                                       1   
created          2019-04-18 05:26:04.446000+00:00   
latitude                                  1.70804   
longitude                                0.870772   
wifi_connection                              True   
carrier_id                                      1   
trans_id          wQMLLmYqiFhSuha9p9B13PMtcyBW_vM   
os_minor                              2.41164e+18   
agent_device                                  NaN   
os_major                              3.90839e+18   
specs_brand                   3576558787748411622   
brand                                         NaN   
timeToClick                               225.311   
touchX           

In [53]:
events.head(5).T

170092   \
index                                             176304   
date                          2019-04-18 00:00:00.027000   
event_id                                              13   
ref_hash                             4314242242878368557   
application_id                                        65   
attributed                                         False   
device_os_version                                    NaN   
device_brand                                 8.64024e+18   
device_model                                 3.07541e+18   
device_city                                          NaN   
session_user_agent                           3.81952e+18   
trans_id                                             NaN   
user_agent                                   8.38837e+18   
event_uuid          eb320d20-1dbe-4240-a948-73e6aedc97c0   
carrier                                              NaN   
kind                                         4.85317e+18   
wifi                                                True   
connection_type                                      NaN   
ip_address                           7661022588781582440   
device_language                              6.97705e+18   
os_es_android                                          1   

                                                 947821   \
index                                             181827   
date                          2019-04-18 00:00:00.071000   
event_id                                             158   
ref_hash                             2068670382964280534   
application_id                                       230   
attributed                                         False   
device_os_version                                    NaN   
device_brand                                         NaN   
device_model                                 6.79488e+18   
device_city                                          NaN   
session_user_agent                           3.81952e+18   
trans_id                                             NaN   
user_agent                                   7.68362e+18   
event_uuid          1d4f4dac-eea4-4767-9256-19b3d2be02e0   
carrier                                              NaN   
kind                                         8.37051e+18   
wifi                                                True   
connection_type                                      NaN   
ip_address                           7503959350608982464   
device_language                              8.44142e+18   
os_es_android                                          0   

                                                 1724915  \
index                                               1159   
date                          2019-04-18 00:00:00.102000   
event_id                                              15   
ref_hash                              726218951290512705   
application_id                                        65   
attributed                                         False   
device_os_version                                    NaN   
device_brand                                 3.08306e+17   
device_model                                 4.06243e+18   
device_city                                          NaN   
session_user_agent                           3.81952e+18   
trans_id                                             NaN   
user_agent                                   4.89308e+18   
event_uuid          0c05bc46-6272-4ad9-9f90-f6c09014ea7f   
carrier                                              NaN   
kind                                         5.50085e+18   
wifi                                                True   
connection_type                                      NaN   
ip_address                           1137401933797700245   
device_language                              6.97705e+18   
os_es_android                                          1   

                                                 2794789  \
index                     

In [54]:
installs.head(5).T

0  \
created                       2019-04-24 06:23:29.495000   
application_id                                         1   
ref_hash                             4716708407362582887   
click_hash                                           NaN   
attributed                                         False   
implicit                                            True   
device_brand                                         NaN   
device_model                                 3.73913e+17   
session_user_agent                            adjust.com   
user_agent                                           NaN   
event_uuid          79837499-2f2a-4605-a663-e322f759424f   
kind                                            app_open   
wifi                                                 NaN   
trans_id                                             NaN   
ip_address                           4243443387795468703   
device_language                              3.30138e+18   
os_es_android                                          0   

                                             1  \
created             2019-04-24 02:06:01.032000   
application_id                               1   
ref_hash                   7143568733100935872   
click_hash                                 NaN   
attributed                               False   
implicit                                 False   
device_brand                               NaN   
device_model                       7.80554e+18   
session_user_agent                  adjust.com   
user_agent                                 NaN   
event_uuid                                 NaN   
kind                                       NaN   
wifi                                       NaN   
trans_id                                   NaN   
ip_address                 4724288679627032761   
device_language                    3.30138e+18   
os_es_android                                0   

                                                       2  \
created                       2019-04-20 10:15:36.274000   
application_id                                         1   
ref_hash                             5230323462636548010   
click_hash                                           NaN   
attributed                                         False   
implicit                                            True   
device_brand                                         NaN   
device_model                                  8.3555e+18   
session_user_agent                            adjust.com   
user_agent                                           NaN   
event_uuid          dda99e3c-9c4b-487d-891c-79f0a02cb4a8   
kind                                            app_open   
wifi                                                 NaN   
trans_id                                             NaN   
ip_address                           8291809486355890410   
device_language                              4.06093e+18   
os_es_android                                          0   

                                                       3  \
created                       2019-04-20 21:56:47.151000   
application_id                                         1   
ref_hash                             5097163995161606833   
click_hash                                           NaN   
attributed                                         False   
implicit                                            True   
device_brand                                         NaN   
device_model                                 2.35577e+18   
session_user_agent                            adjust.com   
user_agent                                           NaN   
event_uuid          7010c3ce-0fcf-46c6-9be8-374cc0e20af4   
kind                                            app_open   
wifi                                                 NaN   
trans_id                                             NaN   
ip_address                           4006811922873399949   
device_language                 

---

# Clicks y su proveniencia desde auctions

Vamos a analizar si hay una relación fuerte entre los clicks registrados y las apariciones en subastas correspondientes cuidando la interpretación ya que no necesariamente todos los clicks deben estar en auctios, pues jampp filtra los datos que procesa y/o nos entrega.

### Vemos la cantidad de "ref_hash" o "device_id" que está en clicks y además en auctios

In [76]:
dispositivos_en_clicks = clicks["ref_hash"].drop_duplicates().to_frame().set_index("ref_hash")

In [78]:
dispositivos_en_clicks["esta_en_auctions"] = dispositivos_en_clicks.index.isin(auctions["device_id"].drop_duplicates())

In [79]:
dispositivos_en_clicks["esta_en_auctions"].value_counts()

True     30873
False      936
Name: esta_en_auctions, dtype: int64

Se observa que 936 dispositivos no aparecen en auctions

### Cantidad de apariciones en ambos datasets

In [81]:
dispositivos_en_clicks["cantidad_apariciones_en_clicks"] = clicks["ref_hash"].value_counts()

In [84]:
dispositivos_en_clicks["cantidad_apariciones_en_auctions"] = auctions.loc[auctions["device_id"].isin(dispositivos_en_clicks.index)]["device_id"].value_counts()

In [116]:
dispositivos_en_clicks.sort_values(by="cantidad_apariciones_en_clicks", ascending=False)

esta_en_auctions  cantidad_apariciones_en_clicks  \
ref_hash                                                                
1999681247947819001              True                              62   
407335392093682287               True                              54   
2093218473313726232              True                              50   
1533680267333692672              True                              43   
1153786347378072057              True                              43   
4053371693190335581              True                              42   
2822277134874324288              True                              40   
406109452263558491               True                              36   
2120207848540337470              True                              36   
1314961964358356814              True                              35   
1555232256616417943              True                              34   
3474363326324590034              True                              33   
6340100001128074332              True                              33   
2191900200262474405              True                              32   
5563201733118730939              True                              32   
336418860500709274               True                              32   
4788456827412910436              True                              32   
3057463431041942739              True                              30   
7059506900875777936              True                              30   
259739936342986289               True                              30   
3012384732267112837              True                              29   
3806738993874726723              True                              29   
5061761822971260588              True                              29   
3870573592046145447              True                              27   
2402572081473431194              True                              27   
8771762948574088406              True                              26   
994952742692964169               True                              26   
2537636611457496026              True                              26   
5242184845582472615              True                              25   
3532045874686965865              True                              25   
...                               ...                             ...   
140383299095162135               True                               1   
1947364434072601155              True                               1   
6644444137177758818              True                               1   
8164825972148734224              True                               1   
401356724880421907               True                               1   
505971290193135133               True                               1   
6770268813951846504              True                               1   
8704594448116561658              True                               1   
4721555596622686776              True                               1   
2184008525713449871              True                               1   
617452377822851611               True                               1   
8636062451117503740              True                               1   
3831401015845936003              True                               1   
1770000139148372020              True                               1   
7058705596667783866              True                               1   
8549915228605648875              True                               1   
1716001858192729347              True                               1   
4871254723127198858              True                               1   
5815272060197798691              True                               1   
2505427418847837541              True                               1   
3129643019616946316              True                               1   
6760365598304701653              True                               1   
280098

### Resumen de cantidad de apariciones en clicks para dispositivos que no estan en auctions

In [87]:
dispositivos_en_clicks.loc[~dispositivos_en_clicks["esta_en_auctions"]]["cantidad_apariciones_en_clicks"].describe()

count    936.000000
mean       1.264957
std        0.761571
min        1.000000
25%        1.000000
50%        1.000000
75%        1.000000
max        8.000000
Name: cantidad_apariciones_en_clicks, dtype: float64

### Miro si puedo asociar datos a cada auction a partir de cada click

In [123]:
dispositivo_en_clicks = clicks.loc[clicks["ref_hash"] == 1999681247947819001]

In [125]:
dispositivo_en_clicks.head()

advertiser_id  action_id  source_id                          created  \
3214              2        NaN          1 2019-04-14 14:50:57.496000+00:00   
3475              2        NaN          2 2019-04-14 05:15:50.055000+00:00   
3494              2        NaN          2 2019-04-14 05:31:53.845000+00:00   
3706              2        NaN          0 2019-04-14 03:59:27.922000+00:00   
3815              2        NaN          8 2019-04-14 16:24:10.980000+00:00   

      latitude  longitude  wifi_connection  carrier_id  \
3214  1.792797   0.866838             True         1.0   
3475  1.714547   0.871535            False         3.0   
3494  1.792518   0.866689             True         7.0   
3706  1.797976   0.874231            False        21.0   
3815  1.767955   0.871124             True         3.0   

                             trans_id      os_minor  agent_device  \
3214  m6Ao7g4-zsfAiMHWz8OuI5gadp02Lss  2.238736e+18           NaN   
3475  Q2XPm9fdOi8QuDm5yUneVtCt9EmesQE  2.238736e+18  6.344087e+18   
3494  6z4F0yxu2INd8eAEiEuVaos-wUxBTAk  2.238736e+18  6.344087e+18   
3706  Ru5MpI7g02MfgThWARC42EXjIaP2rAk  2.238736e+18           NaN   
3815  nEu6uBR-1a7QL4L1d28OGwMGx-etfmo  2.238736e+18           NaN   

          os_major          specs_brand  brand  timeToClick touchX touchY  \
3214  3.581233e+18  3576558787748411622    NaN          NaN    NaN    NaN   
3475  3.581233e+18    71913840936116953    NaN        0.975  0.372  0.680   
3494  3.581233e+18    71913840936116953    NaN        1.075  0.003  0.360   
3706  3.581233e+18    71913840936116953    1.0        0.439  0.753  0.793   
3815  3.581233e+18    71913840936116953    NaN          NaN    NaN    NaN   

                 ref_hash  os_es_android  
3214  1999681247947819001              1  
3475  1999681247947819001              1  
3494  1999681247947819001              1  
3706  1999681247947819001              1  
3815  1999681247947819001              1

In [126]:
dispositivo_en_auctions = auctions.loc[auctions["device_id"] == 1999681247947819001]

In [128]:
dispositivo_en_auctions["ref_type_id"].value_counts()

1    8078
Name: ref_type_id, dtype: int64

---